src/share/classes/com/alibaba/rcm/internal/AbstractResourceContainer.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2020 Alibaba Group Holding Limited. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation. Alibaba designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ */
+
+package com.alibaba.rcm.internal;
+
+import com.alibaba.rcm.Constraint;
+import com.alibaba.rcm.ResourceContainer;
+import sun.misc.JavaLangAccess;
+import sun.misc.SharedSecrets;
+
+import java.util.Collections;
+
+/**
+ * A skeletal implementation of {@link ResourceContainer} that practices
+ * the attach/detach paradigm described in {@link ResourceContainer#run(Runnable)}.
+ * <p>
+ * Each {@code ResourceContainer} implementation must inherit from this class.
+ *
+ * @see ResourceContainer#run(Runnable)
+ */
+
+public abstract class AbstractResourceContainer implements ResourceContainer {
+
+    protected final static JavaLangAccess JLA = SharedSecrets.getJavaLangAccess();
+
+    protected final static AbstractResourceContainer ROOT = new RootContainer();
+
+    public static ResourceContainer root() {
+        return ROOT;
+    }
+
+    public static AbstractResourceContainer current() {
```

Comment:
`current() == null` looks like undefined behavior, we'd better return non-null value, either ROOT or current
```
if (!VM.isBooted()) {
   return root();
}
return JLA.getResourceContainer(Thread.currentThread());
```

Reply:
👍
It's a good idea to ensure that Thread.resourceContainer is always not null. 😋